cpp/src/experimental/louvain.cuh

```diff
@@ -629,591 +292,229 @@ class Louvain {
+        cluster_keys_v_.end(),
+        cluster_weights_v_.data(),
+        d_dst_cluster_cache_,
+        d_dst_cluster_cache_ + dst_cluster_cache_v_.size(),
+        vertex_to_gpu_id_op,
+        stream_);
 
-      cuco::static_map<count_t, count_t> hash_map(
-        capacity, std::numeric_limits<count_t>::max(), count_t{0});
-      detail::create_cuco_pair_t<count_t> create_cuco_pair;
+      map_key_first   = d_dst_cluster_cache_;
+      map_key_last    = d_dst_cluster_cache_ + dst_cluster_cache_v_.size();
+      map_value_first = dst_cluster_weights_v.begin();
+    } else {
+      thrust::transform(rmm::exec_policy(stream_)->on(stream_),
+                        next_cluster_v.begin(),
+                        next_cluster_v.end(),
+                        src_cluster_weights_v.begin(),
+                        [d_cluster_weights = cluster_weights_v_.data(),
+                         d_cluster_keys    = cluster_keys_v_.data(),
+                         num_clusters      = cluster_keys_v_.size()] __device__(vertex_t cluster) {
+                          auto pos = thrust::find(
+                            thrust::seq, d_cluster_keys, d_cluster_keys + num_clusters, cluster);
+                          return d_cluster_weights[pos - d_cluster_keys];
+                        });
+
+      map_key_first   = d_src_cluster_cache_;
+      map_key_last    = d_src_cluster_cache_ + src_cluster_weights_v.size();
+      map_value_first = src_cluster_weights_v.begin();
+    }
 
-      CUDA_TRY(cudaStreamSynchronize(stream_));
+    copy_v_transform_reduce_key_aggregated_out_nbr(
+      handle_,
+      current_graph_view_,
+      thrust::make_zip_iterator(thrust::make_tuple(old_cluster_sum_v.begin(),
+                                                   d_src_vertex_weights_cache_,
+                                                   cluster_subtract_v.begin(),
+                                                   d_src_cluster_cache_,
+                                                   src_cluster_weights_v.begin())),
+
+      d_dst_cluster_cache_,
+      map_key_first,
+      map_key_last,
+      map_value_first,
+      [rank = rank_, total_edge_weight, resolution] __device__(
+        auto src, auto neighbor_cluster, auto new_cluster_sum, auto src_info, auto a_new) {
+        auto old_cluster_sum  = thrust::get<0>(src_info);
+        auto k_k              = thrust::get<1>(src_info);
+        auto cluster_subtract = thrust::get<2>(src_info);
+        auto src_cluster      = thrust::get<3>(src_info);
+        auto a_old            = thrust::get<4>(src_info);
+
+        if (src_cluster == neighbor_cluster) new_cluster_sum -= cluster_subtract;
+
+        weight_t delta_modularity = 2 * (((new_cluster_sum - old_cluster_sum) / total_edge_weight) -
+                                         resolution * (a_new * k_k - a_old * k_k + k_k * k_k) /
+                                           (total_edge_weight * total_edge_weight));
+
+#if 0
```

Comment:
Better delete temporary print code.

Reply:
Thanks.  Hadn't promoted this from draft yet, that was on my todo list.